src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -1350,41 +1351,65 @@ else if (str.length() < 2)
-        } else {
-            digits = value;
-        }
+    private static final Pattern hex = Pattern.compile("\\p{XDigit}+");
+
+    /**
+     * Convert a "#FFF", "#FFFF", "#FFFFFF" or "#FFFFFFFF" hex string to a Color.
+     * If the color specification is bad, an attempt
+     * will be made to fix it up.
+     */
+    static final Color hexToColor(String digits) {
+        int n = digits.length();
+        if (digits.startsWith("#")) {
+            digits = digits.substring(1, Math.min(n, 9));
+            n--;
+        }
+        // CSS Color level 4
+        // - defines color hex code as #[2 digits Red][2 digits Green][2 digits Blue][2 digits Alpha]. With digit 0 ... f.
+        // - allows, webpage passes 3, 4, 6 or 8 digit color code.
+        //   - 3 digits #[R][G][B] ........ represents #[RR][GG][BB]FF
+        //   - 4 digits #[R][G][B][A] ..... represents #[RR][GG][BB][AA]
+        //   - 6 digits #[RR][GG][BB] ..... represents #[RR][GG][BB]FF
+        //   - 8 digits #[RR][GG][BB][AA] . represents #[RR][GG][BB][AA]
+        //
+        // Becareful ! In java.awt.Color hex #[2 digits Alpha][2 digits Red][2 digits Green][2 digits Blue]
+        // Since this method is defined in CSS.java, it must only take in charge CSS Color Level 4 notations.
+        //
+        // According notes below the current OpenJDK implementation is
+        // - 3 digits #[R][G][B]    represents #[RR][GG][BB]FF
+        // - 6 digits #[R][G][B]    represents #[RR][GG][BB]FF
+        //
         // Some webpage passes 3 digit color code as in #fff which is
         // decoded as #000FFF resulting in blue background.
         // As per https://www.w3.org/TR/CSS1/#color-units,
         // The three-digit RGB notation (#rgb) is converted into six-digit form
         // (#rrggbb) by replicating digits, not by adding zeros.
         // This makes sure that white (#ffffff) can be specified with the short notation
         // (#fff) and removes any dependencies on the color depth of the display.
-        if (digits.length() == 3) {
+        if (n == 3 && hex.matcher(digits).matches()) {
             final String r = digits.substring(0, 1);
             final String g = digits.substring(1, 2);
             final String b = digits.substring(2, 3);
-            digits = String.format("%s%s%s%s%s%s", r, r, g, g, b, b);
+            digits = String.format("%s%s%s%s%s%sff", r, r, g, g, b, b);
+        } else if (n == 4 && hex.matcher(digits).matches()) {
+            final String r = digits.substring(0, 1);
+            final String g = digits.substring(1, 2);
+            final String b = digits.substring(2, 3);
+            final String a = digits.substring(3, 4);
+            digits = String.format("%s%s%s%s%s%s%s%s", r, r, g, g, b, b, a, a);
+        } else if (n == 6 && hex.matcher(digits).matches()) {
+            digits = String.format("%sff", digits);
+        } else if (n != 8 || !hex.matcher(digits).matches()) {
+            return null;
         }
```

Comment:
Using `charAt` avoids creating temporary strings of length `1`.

Also, using the `%[argument_index$]conversion` format specifier form allows halving the lengths of the vararg arrays.
```suggestion
        if (n == 3 && hex.matcher(digits).matches()) {
            final char r = digits.charAt(0);
            final char g = digits.charAt(1);
            final char b = digits.charAt(2);
            digits = String.format("%1$s%1$s%2$s%2$s%3$s%3$sff", r, g, b);
        } else if (n == 4 && hex.matcher(digits).matches()) {
            final char r = digits.charAt(0);
            final char g = digits.charAt(1);
            final char b = digits.charAt(2);
            final char a = digits.charAt(3);
            digits = String.format("%1$s%1$s%2$s%2$s%3$s%3$s%4$s%4$s", r, g, b, a);
        } else if (n == 6 && hex.matcher(digits).matches()) {
            digits += "ff";
        } else if (n != 8 || !hex.matcher(digits).matches()) {
            return null;
        }
```

Reply:
@ExE-Boss The GitHub "Commit suggestion" didn't work ! I received an error message. I had to make changes "by hand".

Reply:
@ExE-Boss After a quick performance test, I note that `%1$s%1$s%2$s%2$s%3$s%3$s%4$s%4$s` usage multiplies time execution at least by 4.

I'm going to test if there is other solutions than %n$s or %s. First of all by using a simple concatenation.

Reply:
If allowed `digits = "" + r + r + g + g + b + b + a + a` would be a good candidate.